File: pb/src/main/java/org/sqlbuilder/pb/foreign/VnRole.java
package org.sqlbuilder.pb.foreign;

import org.sqlbuilder.common.NotNull;
import org.sqlbuilder.pb.objects.Theta;

import java.util.Comparator;
import java.util.Objects;

public class VnRole implements Comparable<VnRole>
{
	static public final Comparator<VnRole> COMPARATOR = Comparator.comparing(VnRole::getVnClass).thenComparing(VnRole::getVnTheta);

	private final VnClass vnClass;

	private final Theta vnTheta;

	// C O N S T R U C T O R

	public static VnRole make(final VnClass vnClass, final Theta vnTheta)
	{
		return new VnRole(vnClass, vnTheta);
	}

	private VnRole(final VnClass vnClass, final Theta vnTheta)
	{
		this.vnClass = vnClass;
		this.vnTheta = vnTheta;
	}

	// A C C E S S

	public VnClass getVnClass()
	{
		return this.vnClass;
	}

	public Theta getVnTheta()
	{
		return this.vnTheta;
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		VnRole vnRole = (VnRole) o;
		return vnClass.equals(vnRole.vnClass) && vnTheta.equals(vnRole.vnTheta);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(vnClass, vnTheta);
	}

	// O R D E R I N G

	public int compareTo(@NotNull final VnRole that)
	{
		return COMPARATOR.compare(this, that);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("%s[%s]", this.vnClass, this.vnTheta);
	}
}

File: pb/src/main/java/org/sqlbuilder/pb/foreign/VnRole.kt
package org.sqlbuilder.pb.foreign

import java.util.*

class VnRole private constructor(
    val vnClass: VnClass, val vnTheta: VnLinks,
) : Comparable<VnRole> {

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val vnRole = o as VnRole
        return vnClass == vnRole.vnClass && vnTheta == vnRole.vnTheta
    }

    override fun hashCode(): Int {
        return Objects.hash(vnClass, vnTheta)
    }

    // O R D E R I N G

    override fun compareTo(that: VnRole): Int {
        return COMPARATOR.compare(this, that)
    }

    // T O S T R I N G
    override fun toString(): String {
        return "$vnClass[$vnTheta]"
    }

    companion object {

        val COMPARATOR: Comparator<VnRole> = Comparator
            .comparing<VnRole, VnClass> { it.vnClass }
            .thenComparing { it.vnTheta }

        fun make(vnClass: VnClass, vnTheta: VnLinks): VnRole {
            return VnRole(vnClass, vnTheta)
        }
    }
}
